File: apps/mobile/lib/features/bible/section_title_i18n.dart
/// 段落小标题 zh→en（KJV UI；缺译名回退中文原文）。
library;

import 'bible_book_names.dart';
import 'outlines.dart';

Map<String, String>? _enMap;
Future<void>? _enMapFuture;

void cacheSectionTitleTranslations(Map<String, String> titles) {
  _enMap = titles;
}

Future<void> preloadSectionTitleTranslations(
  Future<Map<String, String>> Function() fetch,
) {
  if (_enMap != null) return Future.value();
  return _enMapFuture ??= fetch().then((titles) {
    cacheSectionTitleTranslations(titles);
  }).catchError((_) {
    cacheSectionTitleTranslations(const {});
  });
}

String sectionTitlesLang(String? versionId) =>
    isEnglishBibleVersion(versionId) ? 'en' : 'zh';

String localizeSectionTitle(String title, {required bool english}) {
  final zh = title.trim();
  if (!english || zh.isEmpty) return zh;
  return _enMap?[zh] ?? zh;
}

List<SectionMark> localizeSectionMarks(
  List<SectionMark> marks, {
  required bool english,
}) {
  if (!english) return marks;
  return marks
      .map(
        (m) => SectionMark(
          m.verse,
          localizeSectionTitle(m.title, english: true),
        ),
      )
      .toList();
}

List<SectionMark> outlineForVersion(
  String bookId,
  int chapter, {
  String? versionId,
}) {
  // NIV 用 API 原生英文标题；本地 outlines 是中文，切勿回落。
  if ((versionId ?? '').trim().toLowerCase() == 'niv') {
    return const [];
  }
  final marks = outlineFor(bookId, chapter);
  return localizeSectionMarks(
    marks,
    english: isEnglishBibleVersion(versionId),
  );
}

File: apps/web/lib/section_title_i18n.ts
/** 段落小标题 zh→en（KJV UI；缺译名回退中文原文）。 */
import type { SectionMark } from './outlines';
import { isEnglishBibleVersion } from './bible_version';

let enMap: Record<string, string> | null = null;
let enMapPromise: Promise<Record<string, string>> | null = null;

export function sectionTitlesLang(versionId: string | null | undefined): 'zh' | 'en' {
  return isEnglishBibleVersion(versionId) ? 'en' : 'zh';
}

async function loadEnMap(): Promise<Record<string, string>> {
  if (enMap) return enMap;
  if (!enMapPromise) {
    enMapPromise = (async () => {
      try {
        const { api } = await import('@/lib/api');
        const data = (await api.sectionTitleTranslations()) as {
          titles?: Record<string, string>;
        };
        enMap = data.titles ?? {};
      } catch {
        enMap = {};
      }
      return enMap;
    })();
  }
  return enMapPromise;
}

export function getSectionTitleEnMapSync(): Record<string, string> | null {
  return enMap;
}

export function preloadSectionTitleTranslations(): void {
  void loadEnMap();
}

export function localizeSectionTitle(
  title: string,
  lang: 'zh' | 'en',
  map?: Record<string, string>,
): string {
  const zh = title.trim();
  if (!zh || lang === 'zh') return zh;
  return map?.[zh] ?? zh;
}

export function localizeSectionMarks(
  marks: SectionMark[],
  lang: 'zh' | 'en',
  map?: Record<string, string>,
): SectionMark[] {
  if (lang === 'zh') return marks;
  return marks.map((m) => ({
    verse: m.verse,
    title: localizeSectionTitle(m.title, lang, map),
  }));
}

export async function localizeSectionMarksAsync(
  marks: SectionMark[],
  lang: 'zh' | 'en',
): Promise<SectionMark[]> {
  if (lang === 'zh') return marks;
  const map = await loadEnMap();
  return localizeSectionMarks(marks, lang, map);
}
